### PR TITLE
Rename manifold methods and attributes

### PR DESCRIPTION
**Pull request by nkoep**

### CODE REVIEW

**Comment by kellertuer** [comment]
This looks super nice!
While there are surely some differences to Manopt.jl also due to the different language, this brings both packages closer in several naming schemes.
Just two small remarks/questions

* For the `tangent_to_ambient` we use `embed(M,p,X)`(and to bring points to the embedding/ambient space `embed(M,p)`, maybe that would be a nice name here, too (where our manifolds are structs and yours are objects, so more like `M.embed`of course)?
* It seems `to_tangent_space` does a projection? Then one could call it `project` as well? (again one would have `M.project(p)`when projecting points and `M.project(p,X)` for tangent projections - or is that too far away from what is usually done in python (in name/function style)?

**Comment by nkoep** [comment]
> This looks super nice! While there are surely some differences to Manopt.jl also due to the different language, this brings both packages closer in several naming schemes. Just two small remarks/questions
> 
>     * For the `tangent_to_ambient` we use `embed(M,p,X)`(and to bring points to the embedding/ambient space `embed(M,p)`, maybe that would be a nice name here, too (where our manifolds are structs and yours are objects, so more like `M.embed`of course)?
> 
>     * It seems `to_tangent_space` does a projection? Then one could call it `project` as well? (again one would have `M.project(p)`when projecting points and `M.project(p,X)` for tangent projections - or is that too far away from what is usually done in python (in name/function style)?

Those are good points, Ronny! I think in light of `projection`, `retraction`, etc. we will go with `embedding` instead of `embed`. So far these methods don't get used much though, only internally in the implementation of two manifolds.

As for `to_tangent_space`, the situation is similar. Right now we only use it in one [place](https://github.com/pymanopt/pymanopt/blob/master/pymanopt/tools/diagnostics.py#L158). The difference with `projection` is that `projection` always assumes the input is a vector in the ambient space, whereas `to_tangent_space` accepts a vector in tangent space format. Seems the main use in manopt at least is in making sure certain vectors that are repeatedly updated in trust-regions and ARC solvers remain in the tangent space, e.g. https://github.com/NicolasBoumal/manopt/blob/master/manopt/solvers/trustregions/tCG.m#L279-L287.

**Comment by kellertuer** [comment]
> I think in light of `projection`, `retraction`, etc. we will go with `embedding` instead of `embed`. So far these methods don't get used much though, only internally in the implementation of two manifolds.

That sounds very reasonable (we have project, retract and embed) if you use nouns, while we use verbs.

> As for `to_tangent_space`, the situation is similar. Right now we only use it in one [place](https://github.com/pymanopt/pymanopt/blob/master/pymanopt/tools/diagnostics.py#L158). The difference with `projection` is that `projection` always assumes the input is a vector in the ambient space, whereas `to_tangent_space` accepts a vector in tangent space format. Seems the main use in manopt at least is in making sure certain vectors that are repeatedly updated in trust-regions and ARC solvers remain in the tangent space, e.g. https://github.com/NicolasBoumal/manopt/blob/master/manopt/solvers/trustregions/tCG.m#L279-L287.

Thanks for the explanation! It seems `to_tangent_space` is really a little different. Our `project(M,p,X)` actually accepts both embedded space format and tangent space format, it projects to the nearest tangent vector and always returns a tangent vector format.

But if one splits both input formats, sure, then the `to_tangent_space` function is merely for numerical stability and could keep its name, since it is also really more like an internal methods.

**Comment by nkoep** [comment]
There's another open point I wanted to ask your opinion on, @kellertuer @NicolasBoumal. The names `ehess` and `hess` never sat quite right with me. In case of `ehess` e.g. for a function from R^n to R we are actually computing a Hessian-vector product rather than the Hessian matrix. This is why I often refer to it by the short form `hvp` instead. However, that name also falls a little flat when we're dealing with functions of matrices or tensors. In that case it's more appropriate to think of `ehess` as applying the linear Hessian operator to a point/vector in the ambient, but I can't think of a good name to accurately describe this fact. Maybe I'm overthinking this, but I want to avoid going from one somewhat misleading name to another.

**Comment by kellertuer** [comment]
For the Hessian in general you can have both representations as a matrix (taking two tangent vectors in a basis representation afterwards to get the value) or as a map from the tangent space to the tangent space (to be precise cotangent space), i.e. it returns a line functional (that together with the other, second vector yields a value).

For the embedding, I do not like, that (a) the embedding is not necessarily clear (which is a generic problem in both the Matlab and Python way it is currently done).

In Julia our hessians have the manifold upfront so we write `H(M, p, X)` returning a tangent vector (so if the second vector on the right of the Hessian matrix is `Y` one still has to compute `inner(M, p, H(M, p, X), Y)`.

To the embedding - since our hessians know their domain, one can either just use the manifold that _is_ the embedding (implicitly, for example `N=Euclidean(3)` when working with `M=Sphere(2)`) or if we have a non-default embedding the manifold M would look like `M=EmbeddedManifold(Euclidean(4), Sphere(2))`(not a useful embedding, the fourth dimension would always be zero, so just for illustration.

The remaining cool stuff python and Manopt can do, Manopt.jl/Manifolds.jl can not yet do with Hessians, but for the gradient we already finished a generic variant, see the second section here https://manoptjl.org/stable/pluto/AutomaticDifferentiation/
(especially `change_representer`). Together with `project` (again), you can have a generic `gradF(N,p)` (in the embedding `N`, which would be `egrad` in your language) and derive the gradient from the restricted function `gradf(M,p)` on the manifold.

I hope to do this generic conversion also with Hessians at some point but did not yet have the time to look at the generic case.

Long story short: We basically do not have `ehess` and for the representation, I think both the matrix is ok (but then you actually need a basis of the tangent space) and the vector one (which does not need a basis).

**Comment by kellertuer** [comment]
I think one could leave out the `riemannian`, since that is clear since it is the gradient for a manifold?

**Comment by nkoep** [comment]
> I think one could leave out the `riemannian`, since that is clear since it is the gradient for a manifold?

That's what I'm currently debating with myself :laughing: I would argue from a user perspective the more natural notion would be the Euclidean gradient. That's what we encourage to provide as well (if any) since the conversion to the Riemannian notion can happen automatically behind the scenes. Since there are two potential gradient arguments to our `Problem` class constructor though I'm thinking it's better to be explicit here too about which gradient in particular we are passing.

**Comment by kellertuer** [comment]
Okay, if you would like to present them on an equal level, then sure, both should get their prefixes.

**Comment by NicolasBoumal** [comment]
> There's another open point I wanted to ask your opinion on, @kellertuer @NicolasBoumal. The names `ehess` and `hess` never sat quite right with me. In case of `ehess` e.g. for a function from R^n to R we are actually computing a Hessian-vector product rather than the Hessian matrix. This is why I often refer to it by the short form `hvp` instead. However, that name also falls a little flat when we're dealing with functions of matrices or tensors. In that case it's more appropriate to think of `ehess` as applying the linear Hessian operator to a point/vector in the ambient, but I can't think of a good name to accurately describe this fact. Maybe I'm overthinking this, but I want to avoid going from one somewhat misleading name to another.

Hello everyone, joining this party a tad late.

The two main reasons hess / ehess do not return matrices in Manopt are as follows:
1. No reasonable algorithm ever needs access to the full Hessian. One always only needs access to Hessian-vector products only. That's the natural way to access a linear map: as its action on vectors.
2. The Hessian is a linear map. To represent it as a matrix, one needs to specify a basis. There is no "natural" basis for a given tangent space of a manifold, hence there is no "natural" matrix representation of the Hessian at a point. One can of course generate a basis (that's rarely cheap) and build a matrix for it (which requires computing very many Hessian-vector products). But that's super expensive and wasteful. Manopt allows it mostly as a convenient tool for debugging / prototyping: see [hessianmatrix](https://www.manopt.org/reference/manopt/tools/hessianmatrix.html).

So to me, it makes sense to call maps which compute hessian-vector products "hess", or "hessian".


About hess / ehess and their names:

1. hess is the Riemannian Hessian. It's not called "rhess" because it's understood that the quantity of reference is the Riemannian one. I could see merits of being explicit, but that's not the choice we made for the Matlab version.
2. ehess is meant to be the "embedded" or "Euclidean Hessian", the precise sense of which is clarified in the documentation only. For the Julia version, the structured use of type inheritance etc. encodes the relation into the code itself, but we don't have that for the Matlab version. The idea is reasonably clear hopefully: a manifold is typically either embedded in another one, or it is a quotient of a total space; and egrad/ehess refer to the gradient and Hessian of the cost function either extended to the embedding space or lifted to the total space. Then Manopt knows how to convert that into the Riemannian quantity.

Admittedly, the name "ehess" is itself somewhat problematic since the embedding space might be another Riemannian manifold rather than a Euclidean space, and the letter "e" makes little sense when what we actually want to refer to is the total space of a quotient. Even worse, for quotient, ehess typically refers to the Euclidean Hessian of a smooth extension of the lifted cost function (think Grassmann as a quotient of Stiefel itself embedded in R^{nxp}). Documentation is supposed to clarify, but it would indeed be nice to have more control.

For the Matlab version, we've been quite intent on preserving backward compatibility. Therefore, big changes in naming conventions are not on the horizon at this time. But I'm still interested in all of use having more extended conversations about the proper way to encode manifolds and their relations to each other. If it becomes clear that we can gain in terms of features / future-proofness by making some changes that break backward-compatibility moderately, that's certainly on the table.

**Comment by nkoep** [comment]
Thanks a lot for the detailed background on the reasoning and naming in Manopt! I can certainly see why "Hessian" is natural to refer to the result of the Hessian operator having acted on a vector at a certain point. While I still would prefer to be more explicit, I really cannot come up with a more concise and descriptive name (given the naming issues with "hvp" I mentioned above). I therefore settled on `{euclidean,riemannian}_{gradient,hessian}` going forward.

The reason to be explicit about what gradient/Hessian we are talking about is that what's natural in one context might not be in another. In the implementation of an optimizer, the Euclidean gradient/Hessian have no real meaning; there we only care about the Riemannian objects. In that context it would be natural to simply call these methods/objects `gradient` and `hessian`, I agree. However, from a user perspective (particularly one who may be coming to the toolbox with little background on the mathematics involved), it may be more natural to refer to the Euclidean notions with `gradient` and `hessian` since that's what most people are familiar with. That's why I decided to be explicit about their names everywhere, most importantly in the user-facing signature of the `Problem` class. Given Pymanopt's emphasis on automatic differentation, the hope is of course that users won't have to deal with gradient/Hessian functions to begin with and simply let the backend take care of computing derivatives. That to me makes it also less problematic that the new names are, admittedly, quite long.

I'm less concerned with the terminology "Euclidean" gradient (e.g.) considering the fact that going up the chain of abstraction, we'll eventually end up in a Euclidean space; either we start with a submanifold readily embedded in Euclidean space, or we consider a quotient manifold, but then as you said the total space (even if another manifold) will be one embedded in Euclidean space. So while maybe not 100 % precise, that terminology feels less worrisome to me.
